onnx/defs/schema.cc

```diff
@@ -49,8 +49,8 @@ const std::string& OpSchema::FormalParameter::GetDescription() const {
   return description_;
 }
 
-bool OpSchema::FormalParameter::IsOptional() const {
-  return is_optional_;
+OpSchema::FormalParameterOption OpSchema::FormalParameter::GetOption() const {
+  return param_option_;
 }
 
 void OpSchema::Verify(const NodeProto& node) const {
```

Comment:
I think that the check if node.input_size() > max_input should be made only if there is no variadic parameter. (Of course, this depends on what max_input is set to for an operator with a variadic parameter.)

Reply:
Exactly on "Of course, this depends on what max_input is set to for an operator with a variadic parameter."

That's why I put a more thought "Remove all "NumInputs", "NumOutputs" API, which seem to be not needed now given we have "optional", "variadic" types specified clearly." with this PR.

I'm thinking of remove the min/max number of input setting. Because, for single/optional input types, it's easy to get to know min/max without an explict set. For variadic, it's always 0-n setting.

Make sense?

If we all agree on removing the "NumInputs, NumOutputs" APIs, I'll do that and change the logic here accordingly too.

Reply:
Yes. I guess others can chime in if they see a need for NumInputs for any other future extension.

Reply:
It might make sense to keep NumInputs just in case we'd need some very fancy operator versions (e.g. for training batchnorm there might be 1 or 3 arguments, but not 2; though it's a bit far-fetched).

But I agree with you that we can do Input() and Output() specs a default. E.g. if none of NumInputs() functions is invoked we just infer it from the Input() spec. Then we can first verify that all existing op specifications are valid before removing NumInput() specs.

But that can happen in a separate PR if your prefer.

Reply:
sure, will make that change in a separate PR.